### PR TITLE
Ensure gap rest consistency  - and several other fixes and improvements for situations with complext track mappings

### PR DESCRIPTION
**Pull request by mike-spa**
Resolves: #21137 

Resolves some long standing inconsistencies that we have with gap rests. There's more work to be done but I think we can be fine with this for now.

Substitute for #23555

Update: my initial solution turned out to be a bit too simplistic. In investigating the problem further, it turned out that we had some serious issues in `Staff::findLinkedStaffInStaff` (which @miiizen had recently improved but not entirely fixed). Namely, the function was working correctly when searching for links in a part starting from the score, but could return completely wrong results when doing the opposite (searching for links in the score starting from a part). I ended up rewriting it, so it now hopefully returns correct results and is clearer too.

### CODE REVIEW

File: src/engraving/dom/edit.cpp [hunk 1]
```diff
@@ -6039,359 +6046,353 @@ void Score::undoAddElement(EngravingItem* element, bool addToLinkedStaves, bool
             ElementType::PEDAL,
             ElementType::LYRICS
         };
-        voice_idx_t voice = track2voice(strack);
-        if (staff->isVoiceVisible(voice) && linkedTracks.empty() && muse::contains(VOICE1_COPY_TYPES, et)) {
```
Comment (cbjeukendrup) [inline src/engraving/dom/edit.cpp:6043]: It looks like `VOICE1_COPY_TYPES` is now unused. The comment above its definition suggests that it might have been important. Are you sure it's not necessary anymore?

File: src/engraving/tests/exchangevoices_data/undoChangeVoice02-ref.mscx [hunk 2]
```diff
@@ -209,6 +209,9 @@
               <tpc>15</tpc>
               </Note>
             </Chord>
+          <location>
+            <fractions>1/4</fractions>
+            </location>
```
Comment (cbjeukendrup) [inline src/engraving/tests/exchangevoices_data/undoChangeVoice02-ref.mscx:214]: Could you explain why these changes occur?

File: src/engraving/dom/edit.cpp (hunk 1 above)
Comment (mike-spa) [reply src/engraving/dom/edit.cpp:6043]: Oops, good catch, thanks

File: src/engraving/tests/exchangevoices_data/undoChangeVoice02-ref.mscx (hunk 2 above)
Comment (mike-spa) [reply src/engraving/tests/exchangevoices_data/undoChangeVoice02-ref.mscx:214]: Those changes occur because of the original commit from the comminity PR. I'll take this as an example. The second voice has 1/4 note and 3/4 of gap rest. 
image
The gap rest isn't written in our files as an actual rest, but as a location move (which I don't like, but that's beside the point). In other words, we represent a 1/4 gap rest as a "move ahead by 1/4" instruction. _Except_, if the destination of the location move coincides with the end of the measure. In that case the location move isn't written because the start of the next measure resets the current tick anyway. 
Now, the difference occurs because previously the gap rest was forced to always be a single rest, which means the underlying gap looks like this 
image
but that gap of 3/4 doesn't get written in the file because the destination coincides with the measure end.
Now, however, we break gaps rests just like normal rests, which means the gap rests now look like this. 
image
Now, the 1/4 rest gets written (that's what creates this diffs in the file), and the 2/4 will be skipped because it's end of the measure.

Comment (cbjeukendrup) [reply src/engraving/tests/exchangevoices_data/undoChangeVoice02-ref.mscx:214]: > _Except_, if the destination of the location move coincides with the end of the measure.

Ah, that explains it! 
Of course, one could have an opinion about how much sense it makes that this causes the quarter rest to be "written" and the half rest not, but I think it doesn't really matter, since the whole concept of gaps and location tags probably won't exist for long anymore anyway.

**Comment by cbjeukendrup** [comment]
One final question: the VTests show some slight spacing difference. Perhaps that's not expected, is it?

**Comment by mike-spa** [comment]
> One final question: the VTests show some slight spacing difference. Perhaps that's not expected, is it?

The horizontal spacing around invisible/hidden/gap rests is quite a complicated business (which also has some other problems of its own, see for instance #19850). The fact that splitting invisible rests can cause small spacing variation is expected, even though undesired. So at this stage I don't think it's a problem.

**Review by cbjeukendrup — APPROVED** [review]

**Review by oktophonie — APPROVED** [review]
